File: Data Structures & Algorithms/count-paths/submission-0.java
class Solution {
    int[][] grid;
    public int uniquePaths(int m, int n) {
        grid = new int[m][n];

        grid[m-1][n-1] = 1;
        return dfs(0, 0, m, n);
    }

    private int dfs(int r, int c, int m, int n) {
        if(r >= m || c >= n) return 0;
        if(grid[r][c] != 0) return grid[r][c];

        return grid[r][c] = dfs(r + 1, c, m, n) + dfs(r, c + 1, m, n);
    }
}

File: Data Structures & Algorithms/count-number-of-islands/submission-0.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        visited = set()
        total = 0

        def dfs(r, c):
            if min(r, c) < 0 or r >= ROWS or c >= COLS or (r, c) in visited or grid[r][c] == "0":
                return
            # if grid[r][c] == 1:
            #     print(visited)
            visited.add((r, c))
            dfs(r + 1, c)
            dfs(r - 1, c)
            dfs(r, c + 1)
            dfs(r, c - 1)
            return

        for row in range(ROWS):
            for col in range(COLS):
                if (row, col) not in visited and grid[row][col] != "0":
                    dfs(row, col)
                    total += 1
        return total
